### PR TITLE
chore: Enable eqeqeq ESLint rule

### PR DESCRIPTION
**Pull request by nwalters512**
This PR is a followup to https://github.com/PrairieLearn/PrairieLearn/pull/4919. It enables the [`eqeqeq`](https://eslint.org/docs/rules/eqeqeq) ESLint rule that forbids the use of `==` except in very specific cases (comparing two literals, evaluating the value of `typeof`, or comparing against `null`).

This was a huge change, as you can see from the diff. I did a lot of the work with good old find-replace, (e.g. I assumed everything that compared with a string literal could trivially safely use `===`, so I replaced all instances of ` == '` with ` === '` and ` != '` with ` !== '), but some required much more manual changes.

Specifically, we actually _rely_ on the type coercion provided by `==` when comparing IDs in many places, since even Postgres sometimes returns them as strings and other times as numbers. The approach I took to fix this was to introduce a new `idsEqual` function that should be used whenever we're comparing IDs. Currently, it's just a wrapper around `==`, but it could be enhanced to do other things in the future. At any rate, I felt that was better than having a bunch of `//eslint-disable` comments scattered everywhere. See [this commit](https://github.com/PrairieLearn/PrairieLearn/commit/954af298a21774cc6f2d72cc88eddcc3dfc5c7ab) to see where I added this function and changed implementations.

### CODE REVIEW

File: docker/start_s3rver.sh
Comment (nwalters512) [inline docker/start_s3rver.sh:17]: I tried to run this with `make start` on my machine outside of docker, but the root of my machine (`/`) is read-only, so this failed. I opted to change this to create the directory in PrairieLearn instead. See the corresponding change to `.gitignore`.

File: lib/id.js
Comment (nwalters512) [inline lib/id.js:29]: Here's the `idsEqual` function I mentioned in the PR description.

**Comment by nwalters512** [comment]
The tests pass, which gives me a reasonable amount of confidence that this is good to go!

File: docker/start_s3rver.sh
Comment (mwest1066) [reply docker/start_s3rver.sh:17]: This seems fine. We were keeping in `/` assuming everyone was running inside a container, but I'm on board with supporting non-containerized dev.

File: lib/id.js
Comment (mwest1066) [reply lib/id.js:29]: Should we add a note about this to https://github.com/PrairieLearn/PrairieLearn/blob/master/docs/dev-guide.md ?

**Comment by mwest1066** [comment]
Everything here makes sense to me.

I'm nervous about this PR because:
1. Are there other cases where we are relying on `==` behavior? The ones that occur to me are `== null` (taken care of), the fact that postgres 64-bit BIGINT maps to string due to Number limits (taken care of), and possible issues with inputs (e.g., form inputs, data from JSON) that we are silently coercing at the moment (I don't know whether there are any).
2. This might cause a lot of merge noise with other outstanding PRs. I don't know how much of an issue this really is.
3. I'm generally a bit nervous of a patch that's so invasive. The risk of weird issues seems high.

But I'm happy to proceed anyway 😄

**Comment by mwest1066** [comment]
> Unrelated, but I think I got through about 20 or so files before realizing how many changes were actually made in this PR 😀

I fortunately looked at the "changed files" count before I started, although I would also have been altered by the fact that my browser slowed down when trying to view the diff 😄

File: lib/editors.js
Comment (jonatanschroeder) [inline lib/editors.js:1182]: I believe these are for QID, so I imagine `===` would be fine...

**Comment by ataraxie** [comment]
I just started watching the PL repo, but I'll feel so free to say my opinion (please just ignore if you think this is no good, haha).

I think this is an important change because `==` is a very common source for bugs. However, I'd also be a bit shaky changing it everywhere all at once in the code base. It might make sense to roll this out a bit more carefully. For example, you could first only replace all instances that are simple string comparisons (not comparing against empty string) or number comparisons (except comparing against 0 and 1). This would be a very safe change. Then, once it's obvious that this didn't break anything (like in a future sprint), you could continue to change another group of checks, etc. But yes, this would require some temporary `//eslint-disable` annotations. But this can be actually kind of nice because it makes it very obvious what still needs to be changed.

P.S. I'm generally on the careful side of things when it comes to these issues and people have often convinced me that something should just be done all-at-once 😅

File: pages/instructorFileTransfer/instructorFileTransfer.js
Comment (nwalters512) [inline pages/instructorFileTransfer/instructorFileTransfer.js:49]:
```suggestion
```

**Review by nwalters512 — COMMENTED** [review]
Comments about this being a bit risky and invasive are completely valid - I agree. To de-risk a bit, I did a self-review of my own changes and found a few things that I'd like a second opinion on. @mwest1066 @jonatanschroeder @nicknytko sorry to ping again, but there's a relatively small set of things that I'd appreciate input on!

**Review by jonatanschroeder — APPROVED** [review]
You pushed a change as I was reviewing, so you may have fixed something already...

File: pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.js
Comment (jonatanschroeder) [reply pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.js:64]: I believe this query allows multiple rows, so `if (result.rowCount > 0)`, or even `if (result.rowCount)`, should be better. That said, this query is weird...

Comment (nwalters512) [reply pages/studentAssessmentInstanceHomework/studentAssessmentInstanceHomework.js:64]: Ahhhh, very good catch. This query is indeed weird, I had to go diving all the way down into `pages/shared/groupWorkInfoContainer.ejs` to see that we do actually expect `group_info` to have multiple rows. I'll update my suggestion.

File: elements/pl-order-blocks/pl-order-blocks.js
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.js:71]: I think using `!==` would work, the intention is to check whether a block is indented by checking the `marginLeft` CSS style is set, as in `var currentIndent = ui.item[0].style.marginLeft;` 10 lines earlier. 

I can't immediately say whether checking `if (currentIndent)` would work, because I don't know too much about Javascript truthy falsy to say if changing the check to `if (currentIndent)` would add any unintentional behaviour (because javascript). Would the javascript wizards of this repo mind chiming in? There might be another much better way to check if the CSS property is set tho.

File: pages/instructorFileTransfer/instructorFileTransfer.js
Comment (nwalters512) [inline pages/instructorFileTransfer/instructorFileTransfer.js:27]:
```suggestion
```

File: pages/instructorFileTransfer/instructorFileTransfer.js
Comment (nwalters512) [reply pages/instructorFileTransfer/instructorFileTransfer.js]: Resolved in e7231f9

**Comment by mwest1066** [comment]
I think that once we move to a staged deployment model (next week) we should go ahead and land this PR. I'll feel confident deploying it if we can try it out with some live data.

**Comment by nwalters512** [comment]
> I think that once we move to a staged deployment model (next week) we should go ahead and land this PR. I'll feel confident deploying it if we can try it out with some live data.

Sounds good - I'll hold off on merging until we get that working!

**Comment by nwalters512** [comment]
Merging master (and thus #5008) into this branch was somewhat of a nightmare. Note to self: don't queue up multiple huge changes like this in the future 🙂 

I'm going to go back and review each and every changed file (again, ugh) to make sure merging occurred without any unintended changes.

**Pull request merged**
